### PR TITLE
Allow `_doc` as a type.

### LINKED ISSUE #27750: Allow `_doc` as a type name
As of 7.0, types support will be gone, the index API will be either `PUT index/_doc/id` or `POST index/_doc` and hits will use `_doc` as a fake type in the response. In order to allow users to have a smoother transition to 7.0, we should allow `_doc` as a type in 6.x.

Relates to #15613

### LINKED ISSUE #27751: Docs should use `_doc` as a type name whenever possible
In order to prepare for the removal of types, we should make docs use `_doc` as a type name whenever possible.

Relates to #15613

### PR DESCRIPTION
**Pull request by jpountz**
Allowing `_doc` as a type will enable users to make the transition to 7.0
smoother since the index APIs will be `PUT index/_doc/id` and `POST index/_doc`.
This also moves same tests and most of the documentation to `_doc` as a type name.

Closes #27750
Closes #27751

### CODE REVIEW

File: docs/reference/mapping/removal_of_types.asciidoc [hunk 1]
```diff
@@ -122,7 +122,7 @@ PUT twitter
         "email": { "type": "keyword" }
       }
     },
-    "tweet": {
+    "_doc": {
```
Comment (clintongormley) [inline docs/reference/mapping/removal_of_types.asciidoc]: this one shouldn't change

File: docs/reference/mapping/removal_of_types.asciidoc [hunk 2]
```diff
@@ -187,7 +187,7 @@ PUT twitter/doc/user-kimchy
```
Comment (clintongormley) [inline docs/reference/mapping/removal_of_types.asciidoc]: `doc` in lines 167 and 180 could become `_doc`

File: docs/reference/mapping/removal_of_types.asciidoc [hunk 3]
```diff
@@ -187,7 +187,7 @@ PUT twitter/doc/user-kimchy
 
 PUT twitter/doc/tweet-1
 {
-  "type": "tweet", <1>
+  "type": "_doc", <1>
```
Comment (clintongormley) [inline docs/reference/mapping/removal_of_types.asciidoc]: this one shouldn't change

File: docs/reference/mapping/removal_of_types.asciidoc [hunk 4]
```diff
@@ -204,7 +204,7 @@ GET twitter/_search
       },
       "filter": {
         "match": {
-          "type": "tweet" <1>
+          "type": "_doc" <1>
```
Comment (clintongormley) [inline docs/reference/mapping/removal_of_types.asciidoc]: this one shouldn't change

File: docs/reference/mapping/removal_of_types.asciidoc [hunk 5]
```diff
@@ -320,7 +320,7 @@ PUT tweets
     "index.mapping.single_type": true
   },
   "mappings": {
```
Comment (clintongormley) [inline docs/reference/mapping/removal_of_types.asciidoc]: In line 242 above, under changes coming in 6.x, mention addition of `_doc` as an acceptable type name to ease the transition to new URLs

File: docs/reference/mapping/removal_of_types.asciidoc [hunk 6]
```diff
@@ -351,7 +351,7 @@ POST _reindex
 {
   "source": {
     "index": "twitter",
-    "type": "tweet"
+    "type": "_doc"
```
Comment (clintongormley) [inline docs/reference/mapping/removal_of_types.asciidoc]: this is the source, so the type name should remain

File: docs/reference/mapping/removal_of_types.asciidoc [hunk 7]
```diff
@@ -351,7 +351,7 @@ POST _reindex
...
   },
```
Comment (clintongormley) [inline docs/reference/mapping/removal_of_types.asciidoc]: in lines 375 and 413 below, use `_doc` instead of `doc`

**Review by clintongormley — CHANGES_REQUESTED** [review]
I couldn't check all 196(!) files, but it looks like you've got most things, thank you. 

Added some comments about the removal of types page.

**Comment by jpountz** [comment]
@clintongormley done

File: docs/reference/mapping/removal_of_types.asciidoc (hunk 2 above)
Comment (clintongormley) [reply docs/reference/mapping/removal_of_types.asciidoc]: Line 167 still needs changing

File: docs/reference/mapping/removal_of_types.asciidoc [hunk 8]
```diff
@@ -244,7 +244,9 @@ Elasticsearch 6.x::
 * Indices created in 5.x will continue to function in 6.x as they did in 5.x.
 
 * Indices created in 6.x only allow a single-type per index.  Any name
-  can be used for the type, but there can be only one.
+  can be used for the type, but there can be only one. The preferred type name
+  is `_doc`, so that index APIs have the same path as they will have in 7.0:
+  `PUT index/_doc/id` and `POST index/_doc`
```
Comment (clintongormley) [inline docs/reference/mapping/removal_of_types.asciidoc]: perhaps `PUT {index}/_doc/{id}` instead?

File: docs/reference/mapping/removal_of_types.asciidoc [hunk 9]
```diff
@@ -257,7 +259,9 @@ Elasticsearch 6.x::
 Elasticsearch 7.x::
 
 * The `type` parameter in URLs are optional.  For instance, indexing
-  a document no longer requires a document `type`.
+  a document no longer requires a document `type`.  The new index APIs
+  are `PUT index/_doc/id` in case of explicit ids and `POST index/_doc`
+  for auto-generated ids.
```
Comment (clintongormley) [inline docs/reference/mapping/removal_of_types.asciidoc]: same here. perhaps `PUT {index}/_doc/{id}` instead?

**Review by clintongormley — APPROVED** [review]
One missed change and a couple of suggestions, otherwise LGTM

**Pull request merged**